l5kit/l5kit/dataset/ego.py

```diff
@@ -98,18 +101,38 @@ def get_frame(self, scene_index: int, state_index: int, track_id: Optional[int]
             )
         data = self.sample_function(state_index, frames, self.dataset.agents, tl_faces, track_id)
 
+        # [history_num_frames + 1, 2]
+        history_positions = np.array(data["history_positions"], dtype=np.float32)
+        # [history_num_frames + 1, 1]
+        history_yaws = np.array(data["history_yaws"], dtype=np.float32)
+
+        # [future_num_frames, 2]
         target_positions = np.array(data["target_positions"], dtype=np.float32)
+        # [future_num_frames, 1]
         target_yaws = np.array(data["target_yaws"], dtype=np.float32)
 
-        history_positions = np.array(data["history_positions"], dtype=np.float32)
-        history_yaws = np.array(data["history_yaws"], dtype=np.float32)
+        # compute estimated target velocities by finite differentiatin on target positions
+        curr_position = np.array([0.0, 0.0])
+        assert np.allclose(history_positions[0], curr_position), "current ego position should be normalized to origin"
+        # [future_num_frames, 2]
+        target_pos_diff = target_positions - np.append([curr_position], target_positions[:-1, :], axis=0)
+        # [future_num_frames,]
+        target_velocities = np.float32(np.linalg.norm(target_pos_diff / self.future_step_time, axis=1))
```

Comment:
this is velocity modules, are we sure we're not interested in the two XY components?

Reply:
shall we return `target_velocities` in xy components and only return `speed` at time 0 (which uses np.linalg.norm on xy components)? the latter (scalalr `speed`) will be used in inference and the former (keeping both xy components) is helpful (e.g. when computing acceleration in x, y components)